[CORE-13280] Adjust synced topic properties

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR makes the following breaking change:

* Renamed `shadowed_topic_properties` to `synced_shadow_topic_properties`

This PR also adds the `exclude_default` flag.  Please refer to `shadow_link.proto` for more details.

This PR now provides the ability to have a default selection of topic properties to replicate.  Additionally it expands the list of unsupported properties.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
cl/model: Defined topic properties in model

Defined the default and required topic properties to sync.  This
selection was done based on discussions with field and engineering to
agree upon a list of topic properties that should be replicated by
default, ones that should never be replicated, and ones that will
always be replicated.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
kafka: Move topic properties to k/protocol

This movies a number of topic property names out of `kafka/server` and
into `kafka/protocol`.  This allows other components to reference these
names without having to depend on `kafka/server`.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added exclude_default

Adds `exclude_default` field to the `topic_metadata_mirroring_config`
model.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Add handling of exclude_default

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Added accessor to get topic properties

This new accessor is used to build a list of topic properties to be
replicated.  This includes the list of properties always replicated,
optionally the default list (if exclude_default is false) and then any
other properties the user wishes to replicate.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Set properties synced status

Provides a list of topic properties that are being replicated.  Depends
on the method defined in the previous commit.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Updates for topic properties

These modifications support future changes for how topic properties are
selected for syncing.  First, the rename is necessary to better
represent what the list indicates.  Finally the new `exclude_default`
field permits users to optional exclude the default topic properties
from being synced and instead rely entirely on
`synced_shadow_topic_properties`

* Renamed `shadowed_topic_properties` to
  `synced_shadow_topic_properties`
* Added status field reporting which properties are being
  synced
* Added `exclude_default` field.  If set then the default topic
  properties will not be included in the list of topic properties that
  are replicated.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/cl: Adjust excluded topic properties

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Support more properties in create_topics

Added missing support for some topic properties in
KafkaCliTools.create_topic.  Also adjusted the `describe_topics_test` to
handle the fact that some configs are now being set on topic creation
making them DYNAMIC.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Add tests for new property behavior

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Fixed issue with rebase

michael-redpanda commented:
Force push:

* Fixed another issue with the rebase

mmaslankaprv commented:
Can you add more detailed commit messages explaining why particular changes are required ?

michael-redpanda committed:
k/server: Make validation_error publicly inherit from std::exception

By default, inheritance is private for classes, meaning this exception
would not be caught even with `catch(const std::exception&)`.  This was
causing some tests to fail because the exception was not caught
properly.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/utils: Clamp all topic durations to max

All referneced calls to parse_and_set_optional_duration set the
clamp_to_duration_max parameter to true.  This is necessary to handle
durations that are larger than the serialiable max duration value.

Signed-off-by: Michael Boquard <michael@redpanda.com>

bharathv commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -229,10 +229,37 @@ message TopicMetadataSyncOptions {
+    // - `max.compaction.lag.ms`
+    //
+    // If this is true, then only the properties listed in
+    // `synced_shadow_topic_properties` will be synced.
+    bool exclude_default = 4;
```
I have a feeling this may not age well.. some user wants to exclude specific defaults (eg: exclude replication factor but not retention and stuff like that). I think to achieve that it requires copying over all default properties into `synced_shadow_topic_properties` before setting this `exclude_default`.

Ideally I think `synced_shadow_topic_properties` should by default populate all default options via `rpk`. wdyt

michael-redpanda replied:
The issue is that we can't have a default list in protobuf.  Setting it via `rpk` is not a bad idea, but would require coordination.

bharathv replied:
Ya exactly, it requires an additional API that `rpk` can get defaults from, I feel thats the cleanest UX here. Obviously needs more work, until that we can just lock the "default list" which are always replicated.. In the next version we can add this additional handling.

michael-redpanda replied:
I think we still need a way for users to select/deselect whether or not they want the default properties.  The UX will be behind RPK/Console.  They'll see a list (generated maybe by docs) and then they can make a decision on what they want to replicate.

bharathv replied:
> select/deselect whether or not they want the default properties.

 For v1 both source and DR should be of same "shape", so that seems like something we can move to v2?

michael-redpanda replied:
I'm of the opinion to allow users to make their own decisions on which topic properties they wish to replicate

bharathv replied:
Agreed on the long term. However, adding this boolean to the proto could be a breaking change if we later decide to go with the approach I proposed above... which is why I was wondering if we could skip it for now (if its ok to in v1)

michael-redpanda replied:
You can mark fields as deprecated in the standard

michael-redpanda replied:
> deprecated (field option): If set to true, indicates that the field is deprecated and should not be used by new code. In most languages this has no actual effect. In Java, this becomes a @Deprecated annotation. For C++, clang-tidy will generate warnings whenever deprecated fields are used. In the future, other language-specific code generators may generate deprecation annotations on the field’s accessors, which will in turn cause a warning to be emitted when compiling code which attempts to use the field. If the field is not used by anyone and you want to prevent new users from using it, consider replacing the field declaration with a [reserved](https://protobuf.dev/programming-guides/proto3/#fieldreserved) statement.

bharathv replied:
umm, I was hoping we could avoid all this dance if we hardcoded default properties for now.. anyway I think its fine if we think there is a proper deprecation path.

michael-redpanda commented:
Force push 1:

* Addressed failures in random ducktape tests
* Improved commit history

Force push 2:

* Fixed lint issue

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
> Can you add more detailed commit messages explaining why particular changes are required ?

Done

michael-redpanda commented:
Force push:

* Addressed test failures

bharathv commented on src/v/cluster/cluster_link/frontend.cc:
```diff
@@ -511,9 +511,9 @@ errc frontend::validate_mutation(const cluster_link_cmd& cmd) const {
-      {"redpanda.remote.readreplica",
-       "redpanda.remote.recovery",
-       "redpanda.remote.allowgaps"}};
+      absl::flat_hash_set<std::string_view>(
+        ::cluster_link::model::disallowed_topic_properties.begin(),
+        ::cluster_link::model::disallowed_topic_properties.end())};
```
nit: I think passing an array is fine, a linear search for this small scale is probably efficient enough.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74092-->
<!--f9031db5-b61a-4b29-a2a5-1b4b37d78c3b-->
<!--c043147b-037b-4729-b17d-f3d557028fdb-->
<!--82b0c593-234c-45e8-bb3a-86707253c0e3-->
<!--feabd52c-ca9e-4eed-8f8a-acc2418ec1e8-->
<!--17f81c53-d2f6-4ae1-b5f9-b57dbd31b4ab-->
<!--1242cb69-3eb2-4132-9894-0371c5c331e1-->
<!--4d7ba2e3-c7bb-484b-b30d-470878c4885e-->
<!--137267fa-0874-4d29-aa94-d7c45f5914eb-->
<!--5b71b37e-3fff-4374-9fde-82ffea39abda-->
<!--cef15752-0efb-4cd8-b4ca-b1bbc85cae73-->
<!--title-->
## Retry command for [Build#74092](https://buildkite.com/redpanda/redpanda/builds/74092)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/describe_topics_test.py::DescribeTopicsTest.test_describe_topics_with_documentation_and_types
tests/rptest/tests/cluster_linking_topic_syncing_test.py::ClusterLinkingTopicSyncingTestNoSecurity.test_topic_syncing_include_exclude
tests/rptest/tests/cluster_linking_topic_syncing_test.py::ClusterLinkingTopicSyncingWithTlsValues.test_topic_syncing_include_exclude
tests/rptest/tests/cluster_linking_topic_syncing_test.py::ClusterLinkingTopicSyncingWithTlsFiles.test_topic_syncing_include_exclude
tests/rptest/tests/cluster_linking_topic_syncing_test.py::ClusterLinkingTopicSyncingWithScram.test_topic_syncing_include_exclude
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--019c3790-939e-449b-8515-21ad6f00cd66-->
<!--TEST RESULTS-->
<!--40f42b6f-eccf-41f3-9399-a20628900c87-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74092">build#74092</a></summary>


| test_class                               | test_method                                       | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                              |
|:-----------------------------------------|:--------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterLinkingTopicSyncingTestNoSecurity | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-890d-44f8-88d2-b48b572b973b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingTestNoSecurity&test_method=test_topic_syncing_include_exclude |
| ClusterLinkingTopicSyncingTestNoSecurity | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec9-3229-47cf-a05e-496f1373ddcd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingTestNoSecurity&test_method=test_topic_syncing_include_exclude |
| ClusterLinkingTopicSyncingWithScram      | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-8911-4b75-89bb-ae18bca15b5e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithScram&test_method=test_topic_syncing_include_exclude      |
| ClusterLinkingTopicSyncingWithScram      | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec9-322d-4e54-9bfe-136bbda9acc7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithScram&test_method=test_topic_syncing_include_exclude      |
| ClusterLinkingTopicSyncingWithTlsFiles   | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-8914-4804-9c80-15f6a3643f00 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsFiles&test_method=test_topic_syncing_include_exclude   |
| ClusterLinkingTopicSyncingWithTlsFiles   | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec9-3232-46a0-8219-072dfa091af0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsFiles&test_method=test_topic_syncing_include_exclude   |
| ClusterLinkingTopicSyncingWithTlsValues  | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-890c-463e-84a1-258ee87ae424 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsValues&test_method=test_topic_syncing_include_exclude  |
| ClusterLinkingTopicSyncingWithTlsValues  | test_topic_syncing_include_exclude                | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec9-3228-4938-a5a2-cf551001e8f2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsValues&test_method=test_topic_syncing_include_exclude  |
| DescribeTopicsTest                       | test_describe_topics_with_documentation_and_types | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-890e-41a6-9146-b33900a4bb90 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DescribeTopicsTest&test_method=test_describe_topics_with_documentation_and_types        |
| DescribeTopicsTest                       | test_describe_topics_with_documentation_and_types | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec9-322a-49eb-a2c9-195e446f376c | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DescribeTopicsTest&test_method=test_describe_topics_with_documentation_and_types        |
| SegmentMsTest                            | test_segment_rolling_with_retention_consumer      | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-8914-4804-9c80-15f6a3643f00 | FLAKY         | 18/21    | upstream reliability is '91.91919191919192'. current run reliability is '85.71428571428571'. drift is 6.20491 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                  |
| RandomNodeOperationsTest                 | test_node_operations                              | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-8913-4ec8-b278-8798242d404b | FLAKY         | 19/21    | upstream reliability is '98.03149606299213'. current run reliability is '90.47619047619048'. drift is 7.55531 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                               |
| WriteCachingFailureInjectionTest         | test_unavoidable_data_loss                        | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/74092#0199dec1-8911-4b75-89bb-ae18bca15b5e | FLAKY         | 18/21    | upstream reliability is '90.39408866995073'. current run reliability is '85.71428571428571'. drift is 4.6798 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss                 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74117">build#74117</a></summary>


| test_class                                         | test_method                                        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:---------------------------------------------------|:---------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cloud_storage/tests/segment_meta_cstore_test | src/v/cloud_storage/tests/segment_meta_cstore_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74117#0199e042-c4c8-442a-a1f7-2302026df9c7 | FAIL          | 0/1      |          |                |
</details>
<!--/body-->

michael-redpanda commented:
Force push:

* Fixed up some tests
